### ISSUE TITLE
Segmentation fault when passing large value for adaptive_avg_pool*

### ISSUE THREAD

**Issue by mijungk** [pytorch/pytorch#42582]
## 🐛 Bug

For `adaptive_avg_pool` `1d`, `2d`, and `3d`, segmentation faults occur when `output_size` is a large value even though a valid tensor for `input` parameter is given. 

## To Reproduce

```python
import torch

# For 1d
torch.nn.functional.adaptive_avg_pool1d(torch.randn([2,2,2]), 9132760301568586890)

# For 2d
torch.nn.functional.adaptive_avg_pool2d(torch.randn([2,2,2,2]), 9132760301568586890)

# For 3d
torch.nn.functional.adaptive_avg_pool3d(torch.randn([2,2,2,2]), 9132760301568586890)

```

## Expected behavior

If `output_size` is not as large as above, the function throws an exception with a proper error message like `RuntimeError: [enforce fail at CPUAllocator.cpp:64] . DefaultCPUAllocator: can't allocate memory`. I would expect such exception with the given value above as well.

## Environment
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.4 LTS
GCC version: Could not collect
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.5.0
[pip] torchvision==0.6.0a0+82fd1c8
[conda] blas 1.0 mkl
[conda] cudatoolkit 10.1.243 h6bb024c_0
[conda] mkl 2020.1 217
[conda] mkl-include 2020.1 217
[conda] mkl-service 2.3.0 py37he904b0f_0
[conda] mkl_fft 1.0.15 py37ha843d7b_0
[conda] mkl_random 1.1.0 py37hd6b4f25_0
[conda] numpy 1.18.1 py37h4f9e942_0
[conda] numpy-base 1.18.1 py37hde5b4d6_1
[conda] pytorch 1.5.0 py3.7_cuda10.1.243_cudnn7.6.3_0 pytorch
[conda] torchvision 0.6.0 py37_cu101 pytorch

cc @ezyang @gchanan @zou3519 @albanD @mruberry

**Comment by thinking-tower** [comment]
Hey @smessmer @dzhulgakov @zou3519, it seems that this issue stems from integer overflow [compute_numel (TensorImpl.h)](https://github.com/pytorch/pytorch/blob/45c5bac87049a9a0593d4d5f060cc8d4b0f83db8/c10/core/TensorImpl.h#L1487) called from [resize_impl_cpu_ (Resize.h)](https://github.com/pytorch/pytorch/blob/45c5bac87049a9a0593d4d5f060cc8d4b0f83db8/aten/src/ATen/native/Resize.h#L32). 

In particular, the following will cause a segmentation fault:
```
import torch
torch.nn.functional.adaptive_avg_pool1d(torch.randn([2,2,2]), ((1 << 63) >> 2)
torch.nn.functional.adaptive_avg_pool2d(torch.randn([2,2,2,2]), ((1 << 31) >> 1)
torch.nn.functional.adaptive_avg_pool3d(torch.randn([4,2,2,2,2]), ((1 << 21) >> 1)
```

whereas this would throw the error as expected:
```
import torch
torch.nn.functional.adaptive_avg_pool1d(torch.randn([2,2,2]), ((1 << 63) - 1) >> 2)
torch.nn.functional.adaptive_avg_pool2d(torch.randn([2,2,2,2]), ((1 << 31) - 1) >> 1)
torch.nn.functional.adaptive_avg_pool3d(torch.randn([4,2,2,2,2]), ((1 << 21) - 1) >> 1)
```

**Comment by kurtamohler** [comment]
Unassigning myself since Mike asked me to stop working on this (https://github.com/pytorch/pytorch/pull/43755#discussion_r482528352)

**Comment by kurtamohler** [comment]
To whomever might pick this issue up in the future, @willwray shared with me a few possible approaches for overflow detection. GCC intrinsics seem like a possible solution, as long as they don't impact performance too much: https://gcc.gnu.org/onlinedocs/gcc/Integer-Overflow-Builtins.html

This benchmark application could be modified to test out the performance of the intrinsics, or any other proposed solution: https://github.com/kurtamohler/pytorch-perf-test-scripts/tree/master/tensor_numel

**Comment by mruberry** [comment]
Thank you, @kurtamohler.

Marking this for retriage. This example seems very adversarial and a fix is nontrivial. I suggest it's not high priority.

One option @gchanan suggests is putting this fix only in debug builds.

**Comment by VitalyFedyunin** [comment]
We consider it rare issue which doesn't deserve high priority label. We are thinking about handling this type of errors in debug builds, to avoid perf hits.

**Comment by DNXie** [comment]
I think this should be fixed by #73719

**Comment by albanD** [comment]
Tentatively closing then. Thanks for pointing it out.

Feel free to re-open if you still see this happening on master.
